releasenotes/notes/remove-symbolic-pulse-subclasses-77314a1654521852.yaml

```diff
@@ -0,0 +1,15 @@
+---
+upgrade:
+  - |
+    Symbolic pulse subclasses :class:`.Gaussian`, :class:`.GaussianSquare`,
+    :class:`.Drag` and :class:`.Constant` have been upgraded to
+    instantiate :class:`SymbolicPulse` rather than subclass itself.
+    All parametric pulse object in pulse programs must be symbolic pulse instance,
+    because subclassing is no longer neccesay. Note that :class:`SymbolicPulse` can
+    uniquely identify a particular envelope with the symbolic expression object
+    defined in :attr:`SymbolicPulse.envelope`.
+  - |
```

Comment:
This is not deprecated yet. I heard we need to prepare alternative before deprecation, then after 1 release we can deprecate it and finally remove. This asks user to replace isinstance with .pulse_type in their code, so I think this is upgrade. Perhaps I'm wrong. Could you please advice? @mtreinish

Reply:
Okay, I see the policy [here](https://qiskit.org/documentation/contributing_to_qiskit.html#deprecation-policy) (maybe you should use `PendingDeprecationWarning` instead of `UserWarning`?). Actually, I didn't understand what the upgrade section meant until I read about it [here](https://docs.openstack.org/reno/latest/user/usage.html) (I was thinking upgraded features rather than suggestions for how to upgrade). So maybe it is standard to describe the upgrade procedure in one release in the upgrade section and then flag the old feature in the deprecation section in the next release.

I wonder about the note above now. that I understand feature vs upgrade better The point about the subclasses now being factories is more of a feature than something requiring action (outside of the instance check described in the second note). Maybe you should emphasize more that SymbolicPulse should not be subclassed and that `Gaussian.__class__.__name__` will stop working because those things do require user action (though I don't think either is common).

Reply:
I agree with @wshanks here, if we're still supporting the old method but adding support for the new factory construction that's a feature note. We should reverse upgrade notes for things where a user would need to take action when upgrading from release N -> N+1. Like when we remove support for the old subclasses (after the deprecation) that would be an upgrade note since the user would have to update their code if they were using the subclasses.